docs/architecture/adr-009-revamp-testkit.md

```diff
@@ -0,0 +1,279 @@
+
+- testing different implementations (traits).
+- succinct tests (useful `util` methods).
+- improving test coverage (Merkle proof generation).
+- integration tests exercising the IBC workflow (relayer-like interface)
+
+## Decision
+
+The main goal of this proposal is to create a test framework that is modular and
+closer to the real blockchain environment. This should also make the existing
+tests succinct and readable. Instead of bootstrapping the mock data, we should
+use valid steps to generate it - so that we know the exact steps to reach a
+state to reproduce in a real environment.
+
+To achieve this, we have broken down the proposal into sub-proposals.
+
+### Adopt a Merkle store for the test framework
+
+The current framework uses `HashMap` and `HashSet` to store data. This works for
+many test scenarios, but it fails to test proof-sensitive scenarios. Because of
+this, we don't have any connection, channel handshake or packet relay tests for
+Tendermint light client.
+
+We generalize
+[`MockContext`](https://github.com/cosmos/ibc-rs/blob/v0.51.0/ibc-testkit/src/testapp/ibc/core/types.rs#L103)
+to use a Merkle store which is used for IBC Context's store. For concrete or
+default implementation, we can use the IAVL Merkle store implementation from
+`informalsystems/basecoin-rs`.
+
+### Modularize the host environment
+
+Currently, we are using `Mock` and `SyntheticTendermint`
+[variants](https://github.com/cosmos/ibc-rs/blob/v0.51.0/ibc-testkit/src/hosts/block.rs#L33-L36)
+as the host environments. To manage these two different environments, we also
+introduced
+[`HostBlocks`](https://github.com/cosmos/ibc-rs/blob/v0.51.0/ibc-testkit/src/hosts/block.rs#L72-75)
+for corresponding host variants.
+
+This creates friction if we have to add a new host variant. It creates the same
+problem as why we have `ibc-derive` crate for `ClientState` and
+`ConsensusState`. This should be refactored by a generic `TestHost` trait that
+maintains its corresponding types e.g. `Block` types, via associated types.
+Finally, we generalize the `MockContext` once more to use this `TestHost` trait
+instead of a concrete enum variant.
+
+This `TestHost` trait should be responsible for generating blocks, headers,
+client and consensus states specific to that host environment.
+
+```rs
+/// TestHost is a trait that defines the interface for a host blockchain.
+pub trait TestHost: Default + Debug + Sized {
+    /// The type of block produced by the host.
+    type Block: TestBlock;
+
+    /// The type of client state produced by the host.
+    type ClientState: Into<AnyClientState> + Debug;
+
+    /// The type of block parameters to produce a block
+    type BlockParams: Debug + Default;
+
+    /// The type of light client parameters to produce a light client state
+    type LightClientParams: Debug + Default;
+
+    /// The history of blocks produced by the host chain.
+    fn history(&self) -> &VecDeque<Self::Block>;
```

Comment:
Is double-ended pushing and popping from the history a common use-case for the block history? In other words, does this need to be a `VecDeque` and not just a regular `Vec`?

Reply:
If we need to test situations where older history needs to be pruned by popping from the front of the `VecDeque`, then I think that is enough to justify using it here.